src/lib.rs

```diff
@@ -280,9 +324,16 @@ impl CodecPipelineImpl {
         let chunk_descriptions =
             self.collect_chunk_descriptions(chunk_descriptions, &output_shape)?;
 
-        py.allow_threads(move || {
-            let codec_options = &self.codec_options;
+        // Adjust the concurrency based on the codec chain and the first chunk description
+        let Some(chunk_descriptions0) = chunk_descriptions.first() else {
+            return Ok(());
+        };
+        let (chunk_concurrent_limit, codec_options) = self.concurrency_chunks_and_codec(
+            chunk_descriptions.len(),
+            chunk_descriptions0.representation(),
+        )?;
```

Comment:
Unless we can come up with a test where we could return `Ok(())` with an empty `chunk_descriptions` i.e., nothing passed into this function, I would be kind of tempted to do

```rust
chunk_descriptions.get_chunk_concurrent_limit_with_codec_options(self)
```
where we implement `get_chunk_concurrent_limit` as part of a trait on the generic return of `collect_chunk_descriptions` so that we can de-duplicate this code.  Seem doable or no?

Reply:
I don't know if zarr-python would ever call this with an empty chunk descriptions anyway, but a user could if they accessed the pipeline directly.

Reply:
True, I like what you did on first glance.  Keeping the check works, yup